pkg/controller/cainjector/controller_test.go

```diff
@@ -0,0 +1,293 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cainjector_test
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	admissionreg "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	certmanager "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	injctrl "github.com/jetstack/cert-manager/pkg/controller/cainjector"
+)
+
+var _ = Describe("Service Controller", func() {
+	var mgrStopCh chan struct{}
+	var mgrDoneCh chan struct{}
+	var client ctrlclient.Client
+
+	var cleanup []runtime.Object
+
+	BeforeEach(func() {
+		By("creating a new manager")
+		mgr, err := ctrl.NewManager(cfg, ctrl.Options{Scheme: scheme})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("creating a client")
+		client, err = ctrlclient.New(cfg, ctrlclient.Options{Scheme: scheme})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("registering the controller with the manager")
+		Expect(injctrl.Register(mgr, injctrl.ValidatingWebhookSetup)).To(Succeed())
+
+		cleanup = nil
+
+		mgrStopCh = make(chan struct{})
+		mgrDoneCh = make(chan struct{})
+		go func() {
+			defer GinkgoRecover()
+			defer close(mgrDoneCh)
+			By("starting the manager")
+			Expect(mgr.Start(mgrStopCh)).To(Succeed())
+		}()
+	})
+
+	AfterEach(func() {
+		By("stopping the manager and waiting to it to finish")
+		if mgrStopCh != nil {
+			close(mgrStopCh)
+			<-mgrDoneCh
+		}
+
+		for _, obj := range cleanup {
+			Expect(client.Delete(context.Background(), obj)).To(Succeed())
+		}
+	})
+
+	Context("for validating webhooks", func() {
+		setupHook := func() (admissionreg.ValidatingWebhookConfiguration, certmanager.Certificate, corev1.Secret) {
+			By("creating a validating webhook pointing to a cert")
+			someURL := "https://localhost:8675"
+			hook := admissionreg.ValidatingWebhookConfiguration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "injected-hook",
+					Annotations: map[string]string{
+						injctrl.WantInjectAnnotation: "default/serving-certs",
+					},
+				},
+				Webhooks: []admissionreg.Webhook{
+					{
+						Name: "hook1.fake.k8s.io",
+						ClientConfig: admissionreg.WebhookClientConfig{
+							URL: &someURL,
+						},
+					},
+					{
+						Name: "hook2.fake.k8s.io",
+						ClientConfig: admissionreg.WebhookClientConfig{
+							Service: &admissionreg.ServiceReference{
+								Name:      "some-svc",
+								Namespace: "default",
+							},
+						},
+					},
+				},
+			}
+			Expect(client.Create(context.Background(), &hook)).To(Succeed())
+			cleanup = append(cleanup, &hook)
+
+			By("creating a certificate marked with the serving annotation")
```

Comment:
This certificate doesn't have any annotations. perhaps "creating a certificate pointed to by annotation"?

Reply:
whoops, name is leftover from old revision.